vulnerabilities/package_managers.py

```diff
@@ -0,0 +1,301 @@
+        for release in json_resp:
+            all_versions.add(release["number"])
+        self.cache[pkg_name] = all_versions
+        return all_versions
+
+
+class NpmVersionAPI(VersionAPI):
+    def get(self, package_name: str) -> Set[str]:
+        """
+        Returns all versions available for a module
+        """
+        package_name = package_name.strip()
+
+        if package_name not in self.cache:
+            releases = set()
+            try:
+                with urlopen(f"https://registry.npmjs.org/{package_name}") as response:
+                    data = json.load(response)
+                    releases = {v for v in data.get("versions", {})}
+            except HTTPError as e:
+                if e.code == 404:
+                    # NPM registry has no data regarding this package, we skip these
+                    pass
+                else:
+                    raise
+
+            self.cache[package_name] = releases
+
+        return self.cache[package_name]
+
+
+class DebianVersionAPI(VersionAPI):
+    async def load_api(self, pkg_set):
+        # Need to set the headers, because the Debian API upgrades
+        # the connection to HTTP 2.0
+        async with ClientSession(
+            raise_for_status=True, headers={"Connection": "keep-alive"}
+        ) as session:
+            await asyncio.gather(
+                *[self.set_api(pkg, session) for pkg in pkg_set if pkg not in self.cache]
+            )
+
+    async def set_api(self, pkg, session, retry_count=5):
+        if pkg in self.cache:
+            return
+        url = "https://sources.debian.org/api/src/{}".format(pkg)
+        try:
+            all_versions = set()
+            response = await session.request(method="GET", url=url)
+            resp_json = await response.json()
+
+            if resp_json.get("error") or not resp_json.get("versions"):
+                self.cache[pkg] = {}
+                return
+            for release in resp_json["versions"]:
+                all_versions.add(release["version"])
+
+            self.cache[pkg] = all_versions
+        # TODO : Handle ServerDisconnectedError by using some sort of
+        # retry mechanism
+        except (ClientResponseError, asyncio.exceptions.TimeoutError, ServerDisconnectedError):
+            self.cache[pkg] = {}
+
+
+class MavenVersionAPI(VersionAPI):
+    def load_to_api(self, pkg_name: str) -> None:
+        if pkg_name in self.cache:
+            return
+
+        artifact_comps = pkg_name.split(":")
+        endpoint = self.artifact_url(artifact_comps)
+        resp = requests.get(endpoint).content
+
+        try:
+            xml_resp = ET.ElementTree(ET.fromstring(resp.decode("utf-8")))
+            self.cache[pkg_name] = self.extract_versions(xml_resp)
+        except ET.ParseError:
+            self.cache[pkg_name] = set()
+
+    @staticmethod
+    def artifact_url(artifact_comps: List[str]) -> str:
+        base_url = "https://repo.maven.apache.org/maven2/{}"
```

Comment:
Maven central canonical URL is usually `https://repo1.maven.org/maven2/` and `https://repo.maven.apache.org/maven2/` is a secondary.

Reply:
Changed. Thanks for the info